File: ts/src/foo.ts
import { IBar, Bar } from './bar'

export class Foo {
    private value: number
    protected bar: IBar

    constructor (value: number) {
        this.value = value
        this.bar = new Bar()
    }

    getTotal() {
        return this.value + this.bar.doSomeThings()
    }
}

export class InternalFoo extends Foo {
    constructor (value: number, bar: IBar) {
        super(value)
        this.bar = bar
    }
}

File: csharp/SomeLibrary/Foo.cs
namespace SomeLibrary
{
    public class Foo
    {
        private IBar bar;
        private int value;

        public Foo(int value)
        {
            this.value = value;
            this.bar = new Bar();
        }

        internal Foo(int value, IBar bar)
        {
            this.value = value;
            this.bar = bar;
        }

        public int GetTotal()
        {
            return value + bar.DoSomeThings();
        }
    }
}
